File: static/sara-voice.js
/**
 * SARA Voice Assistant - FIXED VERSION
 * Fixes:
 * 1. Continuous listening (doesn't stop after first command)
 * 2. Commands recognized on first try
 * 3. Email input works ("dot", "at" conversion)
 * 4. Navigation executes immediately (no need to repeat)
 */

class SARAAssistant {
    constructor() {
        this.isListening = false;
        this.isSpeaking = false;
        
        // Initialize Speech APIs
        const SpeechRecognition = window.SpeechRecognition || window.webkitSpeechRecognition;
        const SpeechSynthesis = window.speechSynthesis;
        
        if (!SpeechRecognition) {
            console.error('❌ Speech Recognition not supported');
            this.recognition = null;
            return;
        }
        
        this.recognition = new SpeechRecognition();
        this.synthesis = SpeechSynthesis;
        
        // CRITICAL: Enable continuous listening
        this.recognition.continuous = true;
        this.recognition.interimResults = true;
        this.recognition.lang = 'en-US';
        
        // Load preferences
        this.voicePreference = localStorage.getItem('sara-voice') || 'professional_female';
        this.blindMode = localStorage.getItem('sara-blind-mode') === 'true';
        
        this.setupHandlers();
        this.setupKeyboardShortcut();
    }
    
    setupHandlers() {
        this.recognition.onstart = () => {
            console.log('✅ Listening started');
            this.isListening = true;
            this.updateStatus('🎤 Listening...');
        };
        
        this.recognition.onresult = (event) => {
            let interim = '';
            let final = '';
            
            for (let i = event.resultIndex; i < event.results.length; i++) {
                const transcript = event.results[i][0].transcript.trim();
                
                if (event.results[i].isFinal) {
                    final += transcript + ' ';
                } else {
                    interim += transcript;
                }
            }
            
            if (interim) {
                this.updateStatus(`Hearing: "${interim}"`);
            }
            
            if (final.trim()) {
                console.log('📍 FINAL:', final.trim());
                this.handleCommand(final.trim());
                this.updateStatus('🎤 Listening for next command...');
            }
        };
        
        this.recognition.onerror = (event) => {
            if (event.error !== 'no-speech' && event.error !== 'network') {
                console.error('Error:', event.error);
            }
        };
        
        this.recognition.onend = () => {
            console.log('Recognition ended');
            this.isListening = false;
            
            // Auto-restart if still enabled
            if (this.shouldKeepListening) {
                setTimeout(() => {
                    try {
                        this.recognition.start();
                    } catch (e) {
                        console.log('Restart:', e.message);
                    }
                }, 500);
            }
        };
    }
    
    setupKeyboardShortcut() {
        document.addEventListener('keydown', (e) => {
            if (e.altKey && e.code === 'Space') {
                e.preventDefault();
                this.toggleListening();
            }
        });
    }
    
    handleCommand(text) {
        console.log('🔍 Processing:', text);
        
        const converted = this.convertText(text);
        const lower = converted.toLowerCase();
        
        console.log('✨ Converted:', converted);
        console.log('📝 Lower:', lower);
        
        // Check for action commands
        if (this.isActionCommand(lower)) {
            console.log('✅ ACTION COMMAND DETECTED');
            this.executeAction(lower);
            return;
        }
        
        // Otherwise form input
        console.log('📋 FORM INPUT DETECTED');
        this.fillForm(text);
    }
    
    convertText(text) {
        let result = text;
        
        // EMAIL: "ashwini dot ingole at gmail dot com" → "ashwini.ingole@gmail.com"
        result = result.replace(/\s+dot\s+/gi, '.');
        result = result.replace(/\s+at\s+/gi, '@');
        
        // NUMBERS
        const numMap = {
            'zero': '0', 'one': '1', 'two': '2', 'three': '3', 'four': '4',
            'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'
        };
        
        for (const [word, num] of Object.entries(numMap)) {
            result = result.replace(new RegExp(`\\b${word}\\b`, 'gi'), num);
        }
        
        return result;
    }
    
    isActionCommand(text) {
        const actions = [
            'login', 'log in', 'sign in', 'signin', 'sign me in', 'log me in',
            'register', 'add achievement', 'add activity',
            'open profile', 'profile', 'portfolio', 'reports', 'notifications',
            'mentor', 'chat', 'dashboard', 'home', 'back', 'go back',
            'dark mode', 'light mode', 'executive', 'submit', 'save',
            'logout', 'exit', 'help', 'stop'
        ];
        
        return actions.some(action => text.includes(action));
    }
    
    executeAction(text) {
        console.log('⚡ EXECUTING:', text);
        
        // LOGIN
        if (this.isLoginCommand(text)) {
            console.log('🔓 Going to login');
            this.speak('Opening login page');
            setTimeout(() => { window.location.href = '/auth/login'; }, 500);
            return;
        }
        
        // REGISTER
        if (text.includes('register')) {
            this.speak('Opening registration');
            setTimeout(() => { window.location.href = '/auth/register'; }, 500);
            return;
        }
        
        // ADD ACHIEVEMENT
        if (text.includes('add achievement')) {
            this.navigate('/student/achievement', 'achievement form');
            return;
        }
        
        // ADD ACTIVITY
        if (text.includes('add activity')) {
            this.navigate('/student/activity', 'activity form');
            return;
        }
        
        // PORTFOLIO
        if (text.includes('portfolio')) {
            this.navigate('/student/portfolio', 'portfolio');
            return;
        }
        
        // REPORTS
        if (text.includes('reports')) {
            this.navigate('/student/reports', 'reports');
            return;
        }
        
        // MENTOR/CHAT
        if (text.includes('mentor') || text.includes('chat')) {
            this.navigate('/student/messages', 'mentor chat');
            return;
        }
        
        // NOTIFICATIONS
        if (text.includes('notifications')) {
            this.navigate('/student/notifications', 'notifications');
            return;
        }
        
        // PROFILE
        if (text.includes('profile')) {
            this.navigate('/student/profile', 'profile');
            return;
        }
        
        // HOME/DASHBOARD
        if (text.includes('home') || text.includes('dashboard')) {
            this.navigate('/student/dashboard', 'dashboard');
            return;
        }
        
        // GO BACK
        if (text.includes('go back') || (text.includes('back') && !text.includes('dark'))) {
            this.speak('Going back');
            window.history.back();
            return;
        }
        
        // DARK MODE
        if (text.includes('dark')) {
            document.body.classList.remove('light-theme', 'executive-theme');
            document.body.classList.add('dark-theme');
            localStorage.setItem('theme', 'dark');
            this.speak('Dark mode enabled');
            return;
        }
        
        // LIGHT MODE
        if (text.includes('light')) {
            document.body.classList.remove('dark-theme', 'executive-theme');
            document.body.classList.add('light-theme');
            localStorage.setItem('theme', 'light');
            this.speak('Light mode enabled');
            return;
        }
        
        // SUBMIT
        if (text.includes('submit')) {
            const btn = document.querySelector('button[type="submit"]');
            if (btn) {
                this.speak('Submitting');
                btn.click();
            }
            return;
        }
        
        // SAVE
        if (text.includes('save')) {
            const btn = document.querySelector('button.save-btn, button[name="save"]');
            if (btn) {
                this.speak('Saving');
                btn.click();
            }
            return;
        }
        
        // LOGOUT
        if (text.includes('logout') || text.includes('exit')) {
            this.speak('Logging out');
            setTimeout(() => { window.location.href = '/auth/logout'; }, 1000);
            return;
        }
        
        // HELP
        if (text.includes('help')) {
            this.showHelp();
            return;
        }
        
        // STOP
        if (text.includes('stop')) {
            this.stopListening();
            return;
        }
    }

    isLoginCommand(text) {
        return /\b(login|log in|logon|log on|sign in|signin|sign me in|log me in|let me in|open login|open sign in|go to login|go to sign in)\b/i.test(text);
    }
    
    navigate(url, name) {
        this.speak(`Opening ${name}`);
        setTimeout(() => { window.location.href = url; }, 500);
    }
    
    fillForm(text) {
        const active = document.activeElement;
        const value = this.valueForField(active, text);
        
        // If input is focused, use it
        if (active && (active.tagName === 'INPUT' || active.tagName === 'TEXTAREA')) {
            active.value = value;
            active.dispatchEvent(new Event('input', { bubbles: true }));
            active.dispatchEvent(new Event('change', { bubbles: true }));
            this.speak(`Set to: ${value}`);
            return;
        }
        
        // Find first empty input
        const inputs = document.querySelectorAll(
            'input:not([type="hidden"]):not([type="submit"]):not([type="button"]), textarea'
        );
        
        for (let input of inputs) {
            if (!input.value || input.value.trim() === '') {
                const fieldValue = this.valueForField(input, text);
                input.focus();
                input.value = fieldValue;
                input.dispatchEvent(new Event('input', { bubbles: true }));
                input.dispatchEvent(new Event('change', { bubbles: true }));
                
                const label = input.placeholder || input.name || 'field';
                this.speak(`Filled ${label}`);
                return;
            }
        }
        
        if (inputs.length > 0) {
            const fieldValue = this.valueForField(inputs[0], text);
            inputs[0].focus();
            inputs[0].value = fieldValue;
            inputs[0].dispatchEvent(new Event('input', { bubbles: true }));
            inputs[0].dispatchEvent(new Event('change', { bubbles: true }));
            this.speak(`Updated field`);
        }
    }

    valueForField(field, text) {
        const raw = String(text || '').trim().replace(/\s+/g, ' ');
        if (!field) return raw;
        const fieldName = `${field.type || ''} ${field.name || ''} ${field.id || ''} ${field.placeholder || ''}`;
        if (/email|e-mail/i.test(fieldName)) return this.convertText(raw).toLowerCase().replace(/\s+/g, '');
        if (/otp|code|pin|mobile|phone|number/i.test(fieldName)) return this.convertText(raw).replace(/\D/g, '');
        return raw;
    }
    
    speak(text) {
        if (!this.synthesis) return;
        
        if (this.isSpeaking) {
            this.synthesis.cancel();
        }
        
        const utterance = new SpeechSynthesisUtterance(text);
        
        const settings = {
            'professional_female': { rate: 1.0, pitch: 1.2 },
            'professional_male': { rate: 1.0, pitch: 0.8 },
            'warm_assistant': { rate: 0.95, pitch: 1.0 },
            'accessibility': { rate: 0.8, pitch: 1.0 }
        }[this.voicePreference] || { rate: 1.0, pitch: 1.0 };
        
        utterance.rate = settings.rate;
        utterance.pitch = settings.pitch;
        utterance.lang = 'en-US';
        
        utterance.onstart = () => { this.isSpeaking = true; };
        utterance.onend = () => { this.isSpeaking = false; };
        
        this.synthesis.speak(utterance);
    }
    
    toggleListening() {
        if (this.isListening) {
            this.stopListening();
        } else {
            this.startListening();
        }
    }
    
    startListening() {
        if (!this.recognition) {
            alert('Speech not supported. Use Chrome, Edge, Safari, or Firefox.');
            return;
        }
        
        try {
            console.log('🎤 Starting');
            this.shouldKeepListening = true;
            if (this.isSpeaking) {
                this.synthesis.cancel();
            }
            this.recognition.start();
            this.updateStatus('🎤 Listening... (Say "login" or your email)');
        } catch (e) {
            console.error('Start error:', e);
        }
    }
    
    stopListening() {
        console.log('🛑 Stopping');
        this.shouldKeepListening = false;
        if (this.recognition && this.isListening) {
            this.recognition.stop();
        }
        this.updateStatus('🎤 Stopped');
    }
    
    updateStatus(message) {
        const status = document.getElementById('sara-status');
        if (status) {
            status.textContent = message;
            status.style.display = 'block';
        }
        console.log('[SARA]', message);
    }
    
    showHelp() {
        const help = `SARA Voice Commands:

LOGIN: "login" or "register"

NAVIGATION:
- "Add Achievement"
- "Add Activity"
- "Open Profile"
- "Portfolio"
- "Reports"
- "Mentor" or "Chat"
- "Notifications"
- "Go Home"
- "Go Back"

FORMS:
- Say email: "ashwini dot ingole at gmail dot com"
- Or any text to fill fields

THEMES:
- "Dark Mode"
- "Light Mode"

ACTIONS:
- "Submit"
- "Save"
- "Logout"`;
        
        alert(help);
    }
}

// Initialize
let sara = null;

document.addEventListener('DOMContentLoaded', () => {
    sara = new SARAAssistant();
    console.log('✅ SARA Voice Ready');
});

// Global functions for widget
window.toggleSARA = () => sara && sara.toggleListening();
window.changeSARAVoice = (voice) => {
    localStorage.setItem('sara-voice', voice);
    if (sara) sara.voicePreference = voice;
};
window.toggleBlindMode = () => {
    const current = localStorage.getItem('sara-blind-mode') === 'true';
    localStorage.setItem('sara-blind-mode', !current);
    if (sara) sara.blindMode = !current;
};

File: app/services/sara_voice.py
"""
SARA - Student Achievement Response Assistant
Voice Assistant Service for SAAMS (Works with manual & voice commands)
Allows blind students to complete entire workflow without keyboard/mouse
"""

from datetime import datetime
from app import db


class VoiceSession(db.Model):
    """Track voice assistant sessions"""
    __tablename__ = 'voice_sessions'
    
    id = db.Column(db.Integer, primary_key=True)
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    session_started = db.Column(db.DateTime, default=datetime.utcnow)
    session_ended = db.Column(db.DateTime)
    total_commands = db.Column(db.Integer, default=0)
    voice_preference = db.Column(db.String(50), default='professional_female')
    blind_mode_enabled = db.Column(db.Boolean, default=False)
    created_at = db.Column(db.DateTime, default=datetime.utcnow)


class SARAVoiceAssistant:
    """
    SARA: Student Achievement Response Assistant
    
    Features:
    - Voice navigation without keyboard/mouse
    - 4 voice options (professional female, professional male, warm, accessibility)
    - Blind student mode with auto-announcements
    - Smart context awareness
    - Voice confirmation for critical actions
    - Session memory
    """
    
    VOICE_OPTIONS = {
        'professional_female': {'name': 'Professional Female', 'speed': 1.0},
        'professional_male': {'name': 'Professional Male', 'speed': 1.0},
        'warm_assistant': {'name': 'Warm Assistant', 'speed': 1.0},
        'accessibility': {'name': 'Accessibility Voice', 'speed': 0.8}
    }
    
    WELCOME_MESSAGES = {
        'startup': "Welcome to SAMS. Would you like to login or register?",
        'login_email': "Please tell me your email address.",
        'login_otp': "OTP has been sent. Please tell me your OTP.",
        'dashboard': "Welcome {name}. You currently have {approved} approved achievements, {pending} pending achievements, and {notifications} new notifications.",
        'blind_mode': "Blind Mode enabled. I will automatically announce available options and important messages.",
    }
    
    COMMAND_RESPONSES = {
        'add_achievement': "Opening Achievement Form. Please tell me the achievement title.",
        'add_activity': "Opening Activity Form. Please tell me the activity name.",
        'open_profile': "Opening your profile.",
        'open_portfolio': "Opening your portfolio.",
        'open_reports': "Opening your reports.",
        'open_mentor_chat': "Opening Mentor Chat. What is the subject?",
        'read_notifications': "Reading your notifications.",
        'logout': "Logging out.",
        'help': "Available commands: Add Achievement, Add Activity, Open Profile, Open Portfolio, Read Notifications, Mentor Chat, Open Reports, Change Theme, Help, and Logout.",
    }
    
    CONFIRMATIONS = {
        'save_achievement': "I heard: {value}. Should I save this?",
        'submit_achievement': "Achievement details recorded. Would you like to upload a certificate?",
        'send_message': "Would you like to send this message to your mentor?",
        'confirm_action': "Should I confirm this action?"
    }
    
    ANNOUNCEMENTS = {
        'page_changed': "You are now on the {page} page.",
        'option_available': "Option: {option}",
        'error': "Error: {message}",
        'success': "Success! {message}",
        'notification': "New notification: {message}",
        'file_selected': "File selected: {filename}",
    }
    
    @staticmethod
    def text_to_speech(text, voice_preference='professional_female', speed=1.0):
        """
        Convert text to speech (frontend handles actual TTS)
        Returns speech instruction for JavaScript
        """
        return {
            'type': 'speak',
            'text': text,
            'voice': voice_preference,
            'speed': speed,
            'timestamp': datetime.utcnow().isoformat()
        }
    
    @staticmethod
    def listen_for_command(timeout=10):
        """
        Listen for voice command (frontend handles actual recording)
        Returns listening instruction for JavaScript
        """
        return {
            'type': 'listen',
            'timeout': timeout,
            'timestamp': datetime.utcnow().isoformat()
        }
    
    @staticmethod
    def process_voice_input(transcribed_text, context='general'):
        """
        Process transcribed voice input
        Converts spoken words to actions
        """
        text = transcribed_text.lower().strip()
        
        # Email conversion: "ashwini dot ingole at gmail dot com" -> "ashwini.ingole@gmail.com"
        if 'dot' in text and 'at' in text:
            text = text.replace(' dot ', '.').replace(' at ', '@')
        
        # Number recognition
        number_words = {
            'zero': '0', 'one': '1', 'two': '2', 'three': '3', 'four': '4',
            'five': '5', 'six': '6', 'seven': '7', 'eight': '8', 'nine': '9'
        }
        for word, digit in number_words.items():
            text = text.replace(word, digit)
        
        return {
            'original': transcribed_text,
            'processed': text,
            'context': context,
            'timestamp': datetime.utcnow().isoformat()
        }
    
    @staticmethod
    def get_dashboard_summary(user):
        """Get summary for dashboard greeting"""
        from app.models import Achievement, Notification
        
        approved = Achievement.query.filter_by(
            student_id=user.id,
            status='Approved'
        ).count()
        
        pending = Achievement.query.filter_by(
            student_id=user.id,
            status='Draft'
        ).count()
        
        notifications = Notification.query.filter_by(
            user_id=user.id,
            is_read=False
        ).count()
        
        return {
            'name': user.full_name.split()[0],  # First name
            'approved': approved,
            'pending': pending,
            'notifications': notifications
        }
    
    @staticmethod
    def get_voice_response(command, user=None, context=None):
        """Get voice response for any command"""
        command = command.lower().strip()
        
        responses = {
            'who are you': "I am SARA, your Student Achievement Response Assistant. I help you manage your achievements and activities without using keyboard or mouse.",
            'good morning': "Good morning! I hope your day is going well. What would you like to do in SAMS today?",
            'hello': "Hello! How can I help you today?",
            'thank you': "You're welcome! If you need help with achievements, activities, reports, or mentor communication, I'm here to assist.",
            'how are you': "I'm doing well and ready to help you manage your achievements and activities. How may I assist you today?",
            'stop': "Listening stopped.",
            'repeat': "Please wait while I repeat the previous message.",
            'help': SARAVoiceAssistant.COMMAND_RESPONSES.get('help', ''),
            'speak slower': "Adjusting speech speed to slower.",
            'speak faster': "Adjusting speech speed to faster.",
            'enable dark mode': "Switching to dark mode.",
            'enable light mode': "Switching to light mode.",
        }
        
        # Check exact matches first
        if command in responses:
            return responses[command]
        
        # Check for command keywords
        for key, response in SARAVoiceAssistant.COMMAND_RESPONSES.items():
            if any(word in command for word in key.split('_')):
                return response
        
        return "I didn't quite understand. Could you please repeat that?"
    
    @staticmethod
    def get_context_announcement(page_name, blind_mode=False):
        """Get announcement when page changes (for blind mode)"""
        if not blind_mode:
            return None
        
        page_announcements = {
            'dashboard': "You are currently on the Dashboard. Available options are: Add Achievement, Add Activity, Profile, Portfolio, Reports, Mentor Chat, Notifications, Settings, and Logout.",
            'achievement': "You are on the Achievement form. Available fields are: Title, Category, Level, Organizer, Event Date, Position, and Description.",
            'activity': "You are on the Activity form. Available fields are: Activity Name, Activity Type, Role, Date, Duration, Organizer, and Description.",
            'profile': "You are on your Profile page. You can view or edit your name, email, phone, department, and role.",
            'portfolio': "You are on your Portfolio page. This shows all your achievements and activities.",
            'reports': "You are on the Reports page. You can view and download your achievement report.",
            'mentor_chat': "You are on Mentor Chat. You can send messages to your mentor and read their replies.",
            'notifications': "You are on Notifications. This shows all your recent updates and mentor messages.",
        }
        
        return page_announcements.get(page_name, f"You are currently on the {page_name} page.")
    
    @staticmethod
    def get_command_options(page_name='dashboard'):
        """Get available voice commands for current page"""
        options = {
            'dashboard': [
                'Add Achievement',
                'Add Activity',
                'Open Profile',
                'Open Portfolio',
                'Open Reports',
                'Open Mentor Chat',
                'Read Notifications',
                'Change Theme',
                'Enable Dark Mode',
                'Enable Light Mode',
                'Help',
                'Logout'
            ],
            'achievement_form': [
                'Submit Achievement',
                'Upload Certificate',
                'Save Draft',
                'Go Back',
                'Help'
            ],
            'mentor_chat': [
                'Send Message',
                'Attach File',
                'Read Latest Reply',
                'Go Back',
                'Help'
            ],
            'notifications': [
                'Read All Notifications',
                'Read Latest Notification',
                'Go Back',
                'Help'
            ]
        }
        
        return options.get(page_name, options['dashboard'])


class VoiceCommandHandler:
    """Handle voice commands and convert to actions"""
    
    @staticmethod
    def parse_command(user_input, user_context=None):
        """Parse user voice input into actionable command"""
        text = user_input.lower().strip()
        
        # Navigation commands
        nav_commands = {
            'add achievement': {'action': 'navigate', 'target': '/student/achievement'},
            'add activity': {'action': 'navigate', 'target': '/student/activity'},
            'open profile': {'action': 'navigate', 'target': '/student/profile'},
            'view profile': {'action': 'navigate', 'target': '/student/profile'},
            'open portfolio': {'action': 'navigate', 'target': '/student/portfolio'},
            'view portfolio': {'action': 'navigate', 'target': '/student/portfolio'},
            'open reports': {'action': 'navigate', 'target': '/student/reports'},
            'view reports': {'action': 'navigate', 'target': '/student/reports'},
            'mentor chat': {'action': 'navigate', 'target': '/student/messages'},
            'open mentor chat': {'action': 'navigate', 'target': '/student/messages'},
            'read notifications': {'action': 'navigate', 'target': '/student/notifications'},
            'notifications': {'action': 'navigate', 'target': '/student/notifications'},
            'open dashboard': {'action': 'navigate', 'target': '/student/dashboard'},
            'go home': {'action': 'navigate', 'target': '/student/dashboard'},
            'go back': {'action': 'navigate', 'target': 'back'},
        }
        
        # Theme commands
        theme_commands = {
            'dark mode': {'action': 'set_theme', 'theme': 'dark'},
            'enable dark mode': {'action': 'set_theme', 'theme': 'dark'},
            'light mode': {'action': 'set_theme', 'theme': 'light'},
            'enable light mode': {'action': 'set_theme', 'theme': 'light'},
            'executive theme': {'action': 'set_theme', 'theme': 'executive'},
            'enable executive theme': {'action': 'set_theme', 'theme': 'executive'},
        }
        
        # Action commands
        action_commands = {
            'logout': {'action': 'logout'},
            'submit': {'action': 'submit_form'},
            'save': {'action': 'save_form'},
            'cancel': {'action': 'cancel_form'},
            'help': {'action': 'show_help'},
            'stop': {'action': 'stop_listening'},
        }
        
        # Check all command types
        for cmd_text, cmd_action in nav_commands.items():
            if cmd_text in text:
                return cmd_action
        
        for cmd_text, cmd_action in theme_commands.items():
            if cmd_text in text:
                return cmd_action
        
        for cmd_text, cmd_action in action_commands.items():
            if cmd_text in text:
                return cmd_action
        
        # If no exact match, return as text input for form fields
        return {
            'action': 'form_input',
            'text': user_input,
            'confidence': 0.85
        }


def create_voice_instructions(action_type, **kwargs):
    """Create instructions for frontend JavaScript to execute"""
    instructions = {
        'speak': {
            'type': 'speak',
            'text': kwargs.get('text', ''),
            'voice': kwargs.get('voice', 'professional_female'),
            'speed': kwargs.get('speed', 1.0)
        },
        'listen': {
            'type': 'listen',
            'timeout': kwargs.get('timeout', 10),
            'max_silence': kwargs.get('max_silence', 2)
        },
        'navigate': {
            'type': 'navigate',
            'url': kwargs.get('url', '/')
        },
        'click': {
            'type': 'click',
            'selector': kwargs.get('selector', '')
        },
        'fill_form': {
            'type': 'fill_form',
            'field': kwargs.get('field', ''),
            'value': kwargs.get('value', '')
        },
        'read_page': {
            'type': 'read_page',
            'elements': kwargs.get('elements', [])
        },
        'confirm': {
            'type': 'confirm',
            'message': kwargs.get('message', 'Is this correct?')
        }
    }
    
    return instructions.get(action_type, {'type': 'unknown'})
